app/src/main/java/mozilla/lockbox/store/DataStore.kt

```diff
@@ -0,0 +1,135 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.lockbox.store
+
+import com.jakewharton.rxrelay2.BehaviorRelay
+import io.reactivex.Observable
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.rxkotlin.addTo
+import io.reactivex.subjects.ReplaySubject
+import mozilla.lockbox.flux.Dispatcher
+import org.mozilla.sync15.logins.LoginsStorage
+import org.mozilla.sync15.logins.MemoryLoginsStorage
+import org.mozilla.sync15.logins.ServerPassword
+import org.mozilla.sync15.logins.SyncResult
+import java.util.*
+
+
+data class DataStoreState(
+        val status: Status,
+        val error: Throwable? = null
+) {
+    enum class Status {
+        UNPREPARED,
+        LOCKED,
+        UNLOCKED,
+        ERRORED
+    }
+}
+
+/**
+ *
+ */
+class DataStore(val dispatcher: Dispatcher = Dispatcher.shared) {
+    internal val compositeDisposable = CompositeDisposable()
+    private val stateSubject: ReplaySubject<DataStoreState> = ReplaySubject.create(1)
+    private val listSubject: BehaviorRelay<List<ServerPassword>> = BehaviorRelay.createDefault(emptyList())
+
+    private val backend: LoginsStorage
+
+    init {
+        // TODO: Replace test data with real backend
+        backend = setupBackend()
+
+        // handle state changes
+        state.subscribe { state ->
+            when (state.status) {
+                DataStoreState.Status.LOCKED -> clearList()
+                DataStoreState.Status.UNLOCKED -> updateList()
+            }
+        }.addTo(compositeDisposable)
+    }
+
+    val state: Observable<DataStoreState> get() = stateSubject
+
+    val list: Observable<List<ServerPassword>> get() = listSubject
+
+    fun unlock() {
+        val doUnlock = { stateSubject.onNext(DataStoreState(DataStoreState.Status.UNLOCKED)) }
+
+        backend.isLocked().whenComplete {
+            if (it) {
+                backend.unlock(getEncryptionKey()).whenComplete {
+                    doUnlock()
+                }
+            }
+        }
+    }
+    fun lock() {
+        val doLock = { stateSubject.onNext(DataStoreState(DataStoreState.Status.LOCKED)) }
+
+        backend.isLocked().whenComplete {
+            if (!it) {
+                backend.lock().whenComplete {
+                    doLock()
+                }
+            }
+        }
+    }
+
+    // backend management
+    private fun setupBackend(): LoginsStorage {
+        val testdata = List<ServerPassword>(10) { createTestItem(it) }
```

Comment:
This seems like a good opportunity to use a Factory.

Reply:
🤔  ... 👍